iroh-memstore/src/lib.rs

```diff
@@ -0,0 +1,247 @@
+                tokio::select! {
+                    biased;
+                    _ = source.next() => {
+                        error!("WatchRequest request received second message");
+                        break;
+                    }
+                    _ = interval.tick() => {
+                        let msg = WatchResponse { version: VERSION.to_string() };
+                        if let Err(_) = sink.send(msg.into()).await {
+                            trace!("WatchRequest response stream closed");
+                        }
+                    }
+                };
+            }
+        });
+        Ok(())
+    }
+
+    async fn handle_version(&self, mut sink: StoreSink, _source: StoreSource) -> Result<()> {
+        // TODO: Generally it would be nice to close the source half before responding.
+        let response = VersionResponse {
+            version: VERSION.to_string(),
+        };
+        sink.send(response.into()).await?;
+        Ok(())
+    }
+
+    async fn handle_put(
+        &mut self,
+        request: PutRequest,
+        mut sink: StoreSink,
+        _source: StoreSource,
+    ) -> Result<()> {
+        let PutRequest { cid, blob, links } = request;
+        let block = StoreBlock { blob, links };
+        self.blocks.insert(cid, block);
+
+        let response = Ok(());
+        sink.send(response.into()).await?;
+        Ok(())
+    }
+
+    async fn handle_put_many(
+        &mut self,
+        request: PutManyRequest,
+        mut sink: StoreSink,
+        _source: StoreSource,
+    ) -> Result<()> {
+        for req in request.blocks {
+            let PutRequest { cid, blob, links } = req;
+            let block = StoreBlock { blob, links };
+            self.blocks.insert(cid, block);
+        }
+
+        let response = Ok(());
+        sink.send(response.into()).await?;
+        Ok(())
+    }
+
+    async fn handle_get(
+        &mut self,
+        request: GetRequest,
+        mut sink: StoreSink,
+        _source: StoreSource,
+    ) -> Result<()> {
+        let block = self.blocks.get(&request.cid);
+
+        let response = Ok(GetResponse {
+            data: block.map(|b| b.blob.clone()),
+        });
+        sink.send(response.into()).await?;
+        Ok(())
+    }
+
+    async fn handle_has(
+        &mut self,
+        request: HasRequest,
+        mut sink: StoreSink,
+        _source: StoreSource,
+    ) -> Result<()> {
+        let has = self.blocks.get(&request.cid).is_some();
+
+        let response = Ok(HasResponse { has });
+        sink.send(response.into()).await?;
+        Ok(())
+    }
+
+    async fn handle_get_links(
+        &mut self,
+        request: GetLinksRequest,
+        mut sink: StoreSink,
+        _source: StoreSource,
+    ) -> Result<()> {
+        let block = self.blocks.get(&request.cid);
+        let links = block.map(|b| b.links.clone());
+
+        let response = Ok(GetLinksResponse { links });
```

Comment:
So what's to stop you from sending a completely inadequate response here? Or send multiple responses in a RPC pattern?

This basically completely ignores all the safety guarantees that the RpcServer gives you. E.g. you redefine the response type of a request, before it would have given you a compile error. Now it gives you nothing.

It might not come as a surprise to you, but I liked the version with the RpcServer better.

Reply:
certainly valid criticism.  i need to think about that.

Reply:
I am not entirely happy with the complexity of the stuff in the giant switch statement. But I also don't like the alternative of having that switch statement generated by a macro. And it does serve a purpose. The stuff in iroh-rpc-types is a protocol defining the return type and interaction pattern for each message. Do a change there, and the compiler will scream at you (albeit with occasionally cryptic errors) until you have updated all client *and* server usages to reflect the change.